### PR TITLE
[FLINK-27303][FLINK-27309] Introduce FlinkConfigManager for efficient config management

### PR DESCRIPTION
**Pull request by gyfora**
This PR improves/reworks the configuration management of the oprator. Problems with the current approach:
  - Configuration and FlinkOperatorConfigurations are passed around everywhere
  - Every component generates effective configurations on their own (without consistency which can cause problems)
  - Configs are generated again and again creating a large number of temporary files in every reconcile iteration
  - Configuration is completely static and impossible to change without operator restart
  
To try to tackle all these problems at once I have introduced the `FlinkConfigManager` object which is responsible for:
 - Generating configuration for deployment / cluster interactions (observe)
 - The config generation logic is consistently aware of currently deployed spec (this is important due to rollback logic)
 - Caching the generated configs for a given specification to speed things up and avoid extra tmp files
 - Allow modifying default configurations on the fly by watching ConfigMap changes
 
While these changes touch almost all components, it is mostly cosmetic. The core improvements are located in the `FlinkConfigManager` class and which config is requested in the different parts of the reconcile logic.

### CODE REVIEW

**Comment by gyfora** [comment]
cc @Aitozi @bgeng777 @morhidi @tweise @wangyang0918

**Comment by wangyang0918** [comment]
Just a quick note: instead of creating a watch on the ConfigMap, I prefer to simply use `ScheduledExecutorService#scheduleAtFixedRate` to detect default configuration file change on the fly. It is similar to `monitorInterval` in log4j. Because naked watch is not stable and could be closed in many cases, we need to use informer instead. But the informer might be an over-kill.

**Comment by gyfora** [comment]
> Just a quick note: instead of creating a watch on the ConfigMap, I prefer to simply use `ScheduledExecutorService#scheduleAtFixedRate` to detect default configuration file change on the fly. It is similar to `monitorInterval` in log4j. Because naked watch is not stable and could be closed in many cases, we need to use informer instead. But the informer might be an over-kill.

Makes sense @wangyang0918 thats a simple change to make it more robust

**Comment by gyfora** [comment]
Reworked the watcher logic and added some more tests

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java [hunk 1]
```diff
@@ -55,31 +54,45 @@
...
+            String namespace,
+            String clusterId,
+            FlinkDeploymentSpec spec,
+            Configuration flinkConfig,
+            boolean forDeployment) {
+        this.namespace = namespace;
+        this.clusterId = clusterId;
         this.spec = spec;
         this.effectiveConfig = new Configuration(flinkConfig);
+        this.forDeployment = forDeployment;
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java:85]: Do we have to add comments for `forDeployment`. It seems mainly to save the create temp template file Right ?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java [hunk 2]
```diff
@@ -177,19 +192,21 @@ public FlinkConfigBuilder applyJobManagerSpec() throws IOException {
...
+    protected FlinkConfigBuilder applyTaskManagerSpec() throws IOException {
         if (spec.getTaskManager() != null) {
             setResource(spec.getTaskManager().getResource(), effectiveConfig, false);
-            setPodTemplate(
-                    spec.getPodTemplate(),
-                    spec.getTaskManager().getPodTemplate(),
-                    effectiveConfig,
-                    false);
+            if (forDeployment) {
+                setPodTemplate(
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java:199]: The temp file may still keep for a long time during the operator running, because it clean up relies on the `deleteOnExit`. Can we expire the temp files when the cache are expired ?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        KubernetesClient kubeClient;
+
+        public ConfigUpdater(KubernetesClient kubeClient) {
+            this.kubeClient = kubeClient;
+        }
+
+        public void run() {
+            try {
+                LOG.debug("Checking for config update changes...");
+                var newConf = readConfiguration(kubeClient);
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:200]: Why not read the config directly from the mounted dir. I think we do not have to interact with the configMap directly. Same at the the `FlinkConfigManager` construct

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java (hunk 2 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java:199]: I will try to add that, good idea

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java (hunk 3 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:200]: Are the mounted files updated automatically if the configmap changes?

Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:200]: As I know, it is

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:200]: nice, that will make this much simpler I agree

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java]: I fixed this and verified that it works correctly

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java:199]: I think over it again, It may bring the sync problem if the config is just used to deploy and the async replace is scheduled. Maybe the cache lifecycle can be align to the job lifecycle, which means we can cleanup the cache and config in the cleanup method ?  WDYT ?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java]: I have added cleanup for the temp files. Also with this I removed the `forDeployment` logic because it is not really necessary anymore (temp files will be cleaned up anyways.)

Set the expiration after access to 1 hour to avoid any problems with cleanup during deployment.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java (hunk 1 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java]: I removed this already

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java (hunk 2 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java]: But you are right this could still be a problem if the clean up is triggered based on the configmap update, I will think about it a little

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java]: @Aitozi I have pushed a fix for this, instead of invalidating the cache I bump a config version long field. This will avoid race conditions between default config updates and temp file cleanup

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                    key.f2,
+                    objectMapper.convertValue(key.f3, FlinkDeploymentSpec.class),
+                    defaultConfig);
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to load configuration", e);
+        }
+    }
+
+    private void scheduleConfigWatcher() {
+        var checkInterval = defaultConfig.get(OPERATOR_DYNAMIC_CONFIG_CHECK_INTERVAL);
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:154]: nit: we can directly use `defaultConfig.get(OPERATOR_DYNAMIC_CONFIG_CHECK_INTERVAL).toMillis();`

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private static final int MAX_CACHE_SIZE = 1000;
+    private static final Duration CACHE_TIMEOUT = Duration.ofMinutes(30);
+
+    private volatile Configuration defaultConfig;
+    private volatile FlinkOperatorConfiguration operatorConfiguration;
+    private final AtomicLong defaultConfigVersion = new AtomicLong(0);
+
+    private final LoadingCache<Tuple4<Long, String, String, ObjectNode>, Configuration> cache;
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:64]: Can we replace the key to a simple inner class instead of `Tuple4`, it may make the meaning more clear

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigBuilder.java]: It looks good now, +1 for this otherwise two other minor comments.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java (hunk 5 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java]: fixed

**Review by Aitozi — APPROVED** [review]
+1

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+/** Configuration manager for the Flink operator. */
+public class FlinkConfigManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkConfigManager.class);
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private static final int MAX_CACHE_SIZE = 1000;
+    private static final Duration CACHE_TIMEOUT = Duration.ofMinutes(30);
+
+    private volatile Configuration defaultConfig;
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:61]: it's the default config onto which the `FlinkDeployment` flinkConfiguration sections are applied.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractDeploymentReconciler.java [hunk 7]
```diff
@@ -39,27 +39,22 @@ public abstract class AbstractDeploymentReconciler implements Reconciler<FlinkDe
...
-        this.operatorConfiguration = operatorConfiguration;
-        this.defaultConfig = defaultConfig;
+        this.configManager = configManager;
     }
 
     @Override
     public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
-        Configuration effectiveConfig = FlinkUtils.getEffectiveConfig(flinkApp, defaultConfig);
-        return shutdownAndDelete(flinkApp, effectiveConfig);
+        return shutdownAndDelete(flinkApp, configManager.getObserveConfig(flinkApp));
```
Comment (bgeng777) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractDeploymentReconciler.java:57]: The older code builds config based on current `flinkApp` but IIUC, here `getObserveConfig`  will use `lastReconciledSpec` or `lastStableSpec`. Is this change on purpose?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractDeploymentReconciler.java:57]: I tried to consistently change it everywhere where we interact with the running job. Because that was technically deployed with either the lastReconciled or lastStable spec as you see it here.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java [hunk 8]
```diff
@@ -119,38 +116,39 @@ public void reconcile(FlinkSessionJob flinkSessionJob, Context context) throws E
                 if (desiredJobState == JobState.RUNNING) {
                     LOG.info("Upgrading/Restarting running job, suspending first...");
                 }
-                stateAfterReconcile = suspendJob(flinkSessionJob, upgradeMode, effectiveConfig);
+                stateAfterReconcile = suspendJob(flinkSessionJob, upgradeMode, deployedConfig);
             }
             if (currentJobState == JobState.SUSPENDED && desiredJobState == JobState.RUNNING) {
                 if (upgradeMode == UpgradeMode.STATELESS) {
-                    submitAndInitStatus(flinkSessionJob, effectiveConfig, null);
+                    submitAndInitStatus(flinkSessionJob, deployedConfig, null);
```
Comment (bgeng777) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java:123]: I think here we should use `getDeployConfig` to submit a new job instead of using `getObserveConfig`?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java:123]: I think this is correct, we need to get the config that is running on the cluster at the moment. Which is the observeConfig
